### DISCUSSION TITLE
Warp on WSL 2

### DISCUSSION THREAD

**Discussion by cadop** [NVIDIA/warp#149] [Q&A]
I am trying to run warp on a wsl2 instance for ubuntu 2204.  

Any ideas?

This is nvidia-smi
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.01              Driver Version: 536.67       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA RTX A4500               On  | 00000000:81:00.0  On |                  Off |
| 30%   39C    P8              29W / 200W |   4338MiB / 20470MiB |      6%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA RTX A4500               On  | 00000000:82:00.0 Off |                  Off |
| 30%   39C    P8              14W / 200W |   4338MiB / 20470MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
```

I can run the pytorch example from https://pytorch.org/tutorials/beginner/pytorch_with_examples.html#pytorch-tensors 

But:
```
import warp as wp
wp.init()

# get all CUDA devices
devices = wp.get_cuda_devices()

>>> Warp CUDA error 100: no CUDA-capable device is detected (/buildAgent/work/3db450722a274445/warp/native/cuda_util.cpp:199)
```

And more generally when I run my code I get:

```
.venv/lib/python3.8/site-packages/warp/context.py", line 2271, in get_current_cuda_device
    raise RuntimeError("CUDA is not available")
RuntimeError: CUDA is not available
```

**Comment by daedalus5** [comment]
This is a full instruction set to get Warp running on WSL 2 from scratch. Perhaps a little more detailed than you need, but in case others need additional info, I'm including all of my steps.

Get Windows Subsystem for Linux

1. Run Windows Terminal as Administrator
2. `wsl --install -d Ubuntu`
3. You can run Ubuntu by searching for Ubuntu.exe from the windows task bar (a system restart may be required first).

Install Git

1. `sudo apt-get install git`
2. `git config --global user.name "Your Name"`
3. `git config --global user.email "youremail@email.com`

Install Git-LFS dependency

1. `curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash`
2. `sudo apt-get install git-lfs`

Install gcc/g++ dependencies

1. `sudo apt update`
2. `sudo apt install build-essential`
3. Default gcc/g++ is version 11, which raises an error while building Warp, so try version 9 instead:
4. `sudo apt -y install gcc-9 g++-9`
5. `sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 9`
6. `sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 9`
7. `gcc --version` should return version 9

Install pip + dependencies

1. `sudo apt install python3-pip`
2. `python3 -m pip install numpy gitpython cmake ninja usd-core`

Install CUDA for WSL

1. Follow instructions [here](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#cuda-support-for-wsl-2)
2. [Install CUDA for WSL](https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&Distribution=WSL-Ubuntu&target_version=2.0&target_type=deb_local)

Build and Install Warp

1. Clone your Warp fork
2. From repo root:
3. `python3 build_lib.py --cuda_path="/usr/local/cuda-12.3"`
4. `sudo python3 -m pip install -e .`


You should now be able to run Warp scripts!

**Reply by cadop** [reply]
Well, i skipped the completely new OS install and it didn't work. If I have time in the future i'll try it again. Otherwise if you have advice on fixing it without new wsl2 instance please let me know!

```
Warp CUDA error 100: no CUDA-capable device is detected (/root/Github/warp/warp/native/cuda_util.cpp:199)
Warp CUDA error 3: initialization error (/root/Github/warp/warp/native/warp.cu:1164)
Warp 1.0.0-beta.2 initialized:
   CUDA devices not available
   Devices:
     "cpu"    | x86_64
   Kernel cache: /root/.cache/warp/1.0.0-beta.2
```

**Reply by daedalus5** [reply]
It's hard for me to say, but for Warp to know where CUDA is, the CUDA toolkit directory on WSL has to be passed to the build step:
`python3 build_lib.py --cuda_path="/usr/local/cuda-12.3"`

If that runs and a CUDA device still cannot be found, there could be an issue if you previously installed a Linux driver within WSL 2 (see link [here](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#cuda-support-for-wsl-2)).
